// ngx-csv-parser/esm2022/lib/ngx-csv-parser.module.mjs
import { NgModule } from '@angular/core';
import { NgxCsvParserComponent } from './ngx-csv-parser.component';
import { NgxCsvParser } from './ngx-csv-parser.service';
import * as i0 from "@angular/core";
class NgxCsvParserModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.0.3", ngImport: i0, type: NgxCsvParserModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.0.3", ngImport: i0, type: NgxCsvParserModule, declarations: [NgxCsvParserComponent], exports: [NgxCsvParserComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.0.3", ngImport: i0, type: NgxCsvParserModule, providers: [NgxCsvParser] }); }
}
export { NgxCsvParserModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.0.3", ngImport: i0, type: NgxCsvParserModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgxCsvParserComponent],
                    imports: [],
                    providers: [NgxCsvParser],
                    exports: [NgxCsvParserComponent]
                }]
        }] });
